### PR TITLE
feat(TUP-23087): Add support for read key from file

### PR DESCRIPTION
**Pull request by fhuaulme**
**What is the problem this Pull Request is trying to solve?**
Studio team may need a way to read a key from the file system. 

**What is the chosen solution to this problem?**
This PR aims to ease this operation with a `KeySource` implementation that reads from file system.

**Important** Undergoing discussion with Studio team (mainly to check if provided code matches all use cases).

**Link to the JIRA issue**
https://jira.talendforge.org/browse/TUP-23087
 
**Please check if the Pull Request fulfills these requirements**
- [x] The PR commit message follows our [guidelines](https://github.com/Talend/daikon/blob/master/CONTRIBUTING.md)
- [x] Tests for the changes have been added (for bug fixes / features, coverage should be over 75% in the new code)
- [x] Docs have been added / updated (for bug fixes / features)
- [ ] Related design / discussions / pages (not in JIRA), if any, are all linked or available in the Pull Request

### CODE REVIEW

File: daikon-crypto/crypto-utils/src/main/java/org/talend/daikon/crypto/KeySources.java [hunk 1]
```diff
@@ -140,4 +144,35 @@ public static KeySource pbkDf2(String password, byte[] salt, int keyLength) {
...
+     * Builds a {@link KeySource} that reads from the given <code>fileName</code>.
+     * <ul>
+     * <li>File is looked up using current thread's classloader (and using
+     * {@link ClassLoader#getResourceAsStream(String)}.</li>
+     * <li>If <code>fileName</code> is not found, a file is created and contains a default key (generated using
+     * <code>defaultKeySource</code> parameter</li>
+     * </ul>
+     * 
+     * @param fileName The file name that contains a key.
+     * @param defaultKeySource A {@link KeySource} to use if file name cannot bbe found.
```
Comment (sgandon) [inline daikon-crypto/crypto-utils/src/main/java/org/talend/daikon/crypto/KeySources.java]:
```suggestion
     * @param defaultKeySource A {@link KeySource} to use if file name cannot be found.
```

File: daikon-crypto/crypto-utils/src/main/java/org/talend/daikon/crypto/KeySources.java [hunk 2]
```diff
@@ -140,4 +144,35 @@ public static KeySource pbkDf2(String password, byte[] salt, int keyLength) {
...
+     * @return A {@link KeySource} that reads key from file or silently generate a file name for next reads.
+     */
+    public static KeySource file(String fileName, KeySource defaultKeySource) {
+        return () -> {
+            final ClassLoader classLoader = Thread.currentThread().getContextClassLoader();
+            try (final InputStream resource = classLoader.getResourceAsStream(fileName)) {
+                if (resource == null) {
+                    final byte[] key = defaultKeySource.getKey();
```
Comment (sgandon) [inline daikon-crypto/crypto-utils/src/main/java/org/talend/daikon/crypto/KeySources.java]: no null check for defaultKeysource

File: daikon-crypto/crypto-utils/src/main/java/org/talend/daikon/crypto/KeySources.java [hunk 3]
```diff
@@ -140,4 +144,35 @@ public static KeySource pbkDf2(String password, byte[] salt, int keyLength) {
...
+     * @param defaultKeySource A {@link KeySource} to use if file name cannot be found.
+     * @return A {@link KeySource} that reads key from file or silently generate a file name for next reads.
+     */
+    public static KeySource file(String fileName, KeySource defaultKeySource) {
+        return () -> {
+            final ClassLoader classLoader = Thread.currentThread().getContextClassLoader();
+            try (final InputStream resource = classLoader.getResourceAsStream(fileName)) {
+                if (resource == null) {
+                    final byte[] key = defaultKeySource.getKey();
+                    try (FileOutputStream fos = new FileOutputStream(fileName)) {
```
Comment (sgandon) [inline daikon-crypto/crypto-utils/src/main/java/org/talend/daikon/crypto/KeySources.java]: why do need to create a file ?

File: daikon-crypto/crypto-utils/src/main/java/org/talend/daikon/crypto/KeySources.java (hunk 2 above)
Comment (fhuaulme) [reply daikon-crypto/crypto-utils/src/main/java/org/talend/daikon/crypto/KeySources.java]: 👍

File: daikon-crypto/crypto-utils/src/main/java/org/talend/daikon/crypto/KeySources.java (hunk 3 above)
Comment (fhuaulme) [reply daikon-crypto/crypto-utils/src/main/java/org/talend/daikon/crypto/KeySources.java]: That's something I need to check with Studio team. I'm fine with throwing a FileNotFoundException if the file is not found (in fact, I'd prefer to do this).

Current code is more a demo than final version though.

**Comment by fhuaulme** [comment]
Will be done in another PR.

**Pull request closed without merging**